src/foundations/type/Type.tsx

```diff
@@ -0,0 +1,124 @@
+import styled, { css } from 'styled-components';
+import { FlattenSimpleInterpolation } from 'styled-components/ts3.6';
+
+import { Utils } from '../../utils';
+
+import { BODY_SIZES, DEFAULT_TYPE_SIZE, HEADING_SIZES, HEADING_ELEMENTS } from './Type.constants';
+import { PType } from './Type.props';
+import { TTypeSize, TTypeType } from './Type.types';
+
+function generateFontStyleVariations(
+    typeSize: TTypeSize,
+    fontType: TTypeType
+): FlattenSimpleInterpolation {
+    return css`
+        &[data-size='${typeSize}'] {
+            font-size: var(--${fontType}-font-size-${typeSize});
+            line-height: var(--${fontType}-line-height-${typeSize});
+
+            &[data-margin='none'] {
+                margin: 0;
+            }
+            &[data-margin='top'] {
+                margin: var(--${fontType}-margin-top-${typeSize}) 0 0;
+            }
+            &[data-margin='bottom'] {
+                margin: 0 0 var(--${fontType}-margin-bottom-${typeSize});
+            }
+            &[data-margin='both'] {
+                margin: var(--${fontType}-margin-top-${typeSize}) 0
+                    var(--${fontType}-margin-bottom-${typeSize});
+            }
```

Comment:
The margin data-attribute styles need to follow the size data attribute in the css selector otherwise the 1000 level margins override all the other levels regardless of what size is selected so I consolidated all the size-related styles into a single function.